Yosemite/Yosemite/Stores/ProductStore.swift

```diff
@@ -323,31 +325,57 @@ private extension ProductStore {
 
     func retrievePopularCachedProducts(siteID: Int64, onCompletion: @escaping ([Product]) -> Void) {
         // Get completed orders
-        let completedOrderPredicate = NSPredicate(format: "statusKey ==[c] %@", OrderStatusEnum.completed.rawValue)
-        let sitePredicate = NSPredicate(format: "siteID == %lld", siteID)
-        let predicate = NSCompoundPredicate(andPredicateWithSubpredicates: [completedOrderPredicate, sitePredicate])
         let completedStorageOrders = sharedDerivedStorage.allObjects(ofType: StorageOrder.self,
-                                                      matching: predicate,
+                                                      matching: completedOrdersPredicate(from: siteID),
                                                       sortedBy: nil)
 
         let completedOrders = completedStorageOrders.map { $0.toReadOnly() }
 
         // Get product ids sorted by occurence
         let completedOrdersItems = completedOrders.flatMap { $0.items }
         let productIDCountDictionary = completedOrdersItems.reduce(into: [:]) { counts, orderItem in counts[orderItem.productID, default: 0] += 1 }
-        let sortedByOccurenceProductIDs = productIDCountDictionary.sorted { $0.value > $1.value }.map { $0.key }
+        let sortedByOccurenceProductIDs = productIDCountDictionary
+            .sorted { $0.value > $1.value }
+            .map { $0.key }
+            .uniqued()
 
         // Retrieve products from product ids and finish
-        let products = sortedByOccurenceProductIDs
+        let products = products(from: sortedByOccurenceProductIDs)
+        onCompletion(products)
+    }
+
+    func retrieveRecentlySoldCachedProducts(siteID: Int64, onCompletion: @escaping ([Product]) -> Void) {
+        let completedStorageOrders = sharedDerivedStorage.allObjects(ofType: StorageOrder.self,
+                                                      matching: completedOrdersPredicate(from: siteID),
+                                                      sortedBy: [NSSortDescriptor(key: #keyPath(StorageOrder.datePaid), ascending: false)])
```

Comment:
Just a side-note: There's a gotcha with the combination between `datePaid` and the `complete` order status that I believe is no concern here, but is something to be aware of:

`datePaid` could have a value on unpaid/non-completed orders if a merchant has switched order status. For example switching an order to be "completed", then switch it back to be "pending" will result on an unpaid order with a datePaid value.

Reply:
Thanks for the information!